src/modm/platform/clock/samg/module.lb

```diff
@@ -15,7 +15,7 @@ def init(module):
     module.description = "Clock Generator (CKGR)"
 
 def prepare(module, options):
-    if not options[":target"].has_driver("pmc:samg*"):
```

Comment:
I think this needs to call out samg and samv specifically, because this driver will not work with, e.g. samd parts.

Reply:
Oh, yes, absolutely. In the current state this commit is just a hack put together to make this MCU booting and taking over gpio. So far I’m having trouble with that part.

Reply:
Fixed - is this what you'd expect? SAMV driver is named just "sam".

Also, organizational question here: right now the SAMG/SAMV driver is in "samg" folder and SAML/SAMD driver in "sam" folder. Should we rename something here?

Reply:
I'll let @salkinium comment on what to actually do here. But I'll write some comments since I've been looking at it recently:

The clock setup is similar between samg/samv at a glance, but I'm seeing several differences: 
- PLLB is renamed to USB PLL
- The PLL input comes from main clock instead of slow clock
- There's no PLLADIV2 bit  to divide the PLLA output by 2 in the PMC_MCKR register. 

I bet there are some other small things that will pop up as you get into it.

Peripheral clock control looks to be indentical; at least I haven't noticed a difference yet. 

For the ClockGen API, I basically assumed that it was OK to require that PLLA is used for master clock, and PLLB is used for USB clock. The SAMV70 just put this requirement into hardware.

I'm not sure what to call a driver that supports samg and samv but not samd. Perhaps "clocks/sam_pmc", with the intent that it cover all SAM series that have the "PMC" module (samd does not).

Reply:
Don't overthink it, "clocks/sam_pmc" is perfectly fine. Remember that lbuild normalizes the different implementation anyways via the module name `modm:platform:clockgen` so that the outwards facing API is the same or at least similiar.

Reply:
> Don't overthink it, "clocks/sam_pmc" is perfectly fine. 

I appreciate your aversion to getting too caught up in the silly things! :)

> Remember that lbuild normalizes the different implementation anyways via the module name `modm:platform:clockgen` so that the outwards facing API is the same or at least similiar.

I fear we may have missed the mark on this, if the intent was a standard-ish clock setup naming scheme. Not that I'm super bothered by this. I feel like clock setup is one of those things that's  just not a good candidate for wide/uniform  cross-chip support.

samg: :platform:clockgen -> class ClockGen
sam: platform:gclk -> GenericClockController
stm32: :platform:rcc -> Rcc (with a deprecated alias to ClockControl)

I chose ClockGen because "Clock Generator" is the datasheet name, and it seemed reasonable (Generic Clock Controller can be found in the SAMD datasheets, I think).

Reply:
Ah right, I thought SAMD and SAMG were similiar. If no consistent naming can be found, then the peripheral name as is shown in the datasheet should be used, so we should keep that naming here.

Unfortunately finding a consistent naming that remains consistent for future ports is a bit like reading tea leaves and it doesn't help that the vendors introduce additional inaccuracies in their own documentation. I'm a bit unhappy about all of this, but I don't know any better approach that isn't confusing somewhere…

Reply:
Also sorry this is so complicated, I will prioritize adding a Readme to the platform folder, describing these porting issues in more detail.

Reply:
I tidied up the clock hack, now it shouldn't break samg anymore. It's still WIP, but PTAL at `Implement basic clock support for SAMV` commit to check if it's going in good direction.

Questions & notes:
- Should we rename some samg/samv identifiers to have the same API on both targets (e.g. UPLL vs PLLB) or stick with datasheet naming
- I see you didn't pull in all peripheral identifiers into ClockPeripheral - how did you choose them?
- There's no `PLLxDIV2`, but there is `CKGR_PLLAR_DIVA`
- I'm not sure if PRES/CSS order applies for SAMV as well or it's just SAMG quirk

Reply:
> I tidied up the clock hack, now it shouldn't break samg anymore. It's still WIP, but PTAL at `Implement basic clock support for SAMV` commit to check if it's going in good direction.
> 
> Questions & notes:
> 
> * Should we rename some samg/samv identifiers to have the same API on both targets (e.g. UPLL vs PLLB) or stick with datasheet naming

I think it's usually better to match the datasheet naming, even though it makes the implementations less similar.

We could provide higher level configuration functions, like `enableUsb48MhzClk()` that work  for both, although maybe this is best left in the board definitions. There isn't a single way to configure this, e.g. you can share PLLA for full-speed USB if you set the main clk up right (e.g. run the processor at 96M and div/2 for USB), or use the UPLL for USB. For high-speed USB it looks like you have to use the UPLL. For samv, the input clock depends on the crystal installed, for samg it can only run off of the 32kHz so you know the input clock. 

> * I see you didn't pull in all peripheral identifiers into ClockPeripheral - how did you choose them?

Did I miss some? I intended to pull all of the periphs with IDs defined in samg55j19.h. However, I expect your processor will have a different set of periph IDs. I don't actually know how to extract these automatically; AFAIK they aren't in the device definitions, and just have to be manually provided.

> * I'm not sure if PRES/CSS order applies for SAMV as well or it's just SAMG quirk

Section 31.17 of the SAMV70 datasheet calls it out as well.

Reply:
Some of the missing IDs from `samg55j19.h` are `ID_SUPC`, `ID_RSTC`, `ID_RTC`, just to name a few. Are they not configurable?

I pushed next revision with all the todos (except peripheral IDs) fixed. More observations:
- I removed uPllFrequency getter since it's always 480MHz
- I removed possibility to configure main clock on top of UPLL (it doesn't make sense, does it?)
- I implemented enableUPll, but I have no way to test it, sorry
- You noticed V70 runs off of MAIN_CLK, not SLOW_CLK. I think this might be the reason why my LEDs blink so slowly

Reply:
> Some of the missing IDs from `samg55j19.h` are `ID_SUPC`, `ID_RSTC`, `ID_RTC`, just to name a few. Are they not configurable?

Oh yeah. These don't have a configurable clock; i.e. no bit in the PCERx registers.
> 
> I pushed next revision with all the todos (except peripheral IDs) fixed. More observations:
> 
> * I removed possibility to configure main clock on top of UPLL (it doesn't make sense, does it?)

If you mean including  UPLL in the MasterClkSource enum, I don't see any reason not to. It's one of the options, and if the UPLL is engaged generating a 480MHz clock for the USB, one may want to use it div-by-2 to run the core at 240MHz. It saves some power to not run the second PLL. I'm not at all suggesting you should aim to support everything possible here; but in this case it's pretty trivial to include all four mux options in the enum. Seems worth it to me.

Reply:
Right, I just checked in Atmel START and indeed there is an option to select UDPLL as clock source for MCC. But for now I need to make the basic config working...

Reply:
I think with the most recent version of this PR I incorporated all of the feedback from this thread. Please LMK if I missed anything!